src/server/frontend_wayland/window_wl_surface_role.cpp

```diff
@@ -468,26 +471,44 @@ void mf::WindowWlSurfaceRole::create_scene_surface()
         observer->attrib_changed(scene_surface.get(), mir_window_attrib_focus, focus_state);
     }
 
-    // Send wl_surface.enter events for every output
-    // TODO: send enter/leave when the surface actually enters and leaves outputs
-    output_manager->current_config().for_each_output([&](graphics::DisplayConfigurationOutput const& conf)
-        {
-            auto const output = output_manager->output_for(conf.id);
-            if (output)
-            {
-                output.value()->for_each_output_bound_by(
-                    client,
-                    [&](OutputInstance* output)
-                    {
-                        surface.value().send_enter_event(output->resource);
-                    });
-            }
-        });
-
     // Invalidates mods
     pending_changes.reset();
 }
 
+void mf::WindowWlSurfaceRole::handle_enter_output(graphics::DisplayConfigurationOutputId id)
+{
+    if (surface)
+    {
+        output_manager->for_each_output_bound_by(
+            client,
+            [&](OutputInstance* output)
+            {
+                auto const& config{output->global.value().current_config()};
+                if (config.valid() && config.id == id)
+                {
+                    surface.value().send_enter_event(output->resource);
+                }
+            });
```

Comment:
```suggestion
        if (auto const output = output_manager->output_for(id))
        {
            surface.value().send_enter_event((*output)->resource);
        }
```

Reply:
We need `OutputInstance` to access the resource.